### PR TITLE
Support 45 Mhz clock and new ADC readout protocol

### PR DESCRIPTION
**Pull request by SkyToGround**
### Issue reference / description

The branch you merge from should already reference an event-formation-unit github ticket number. You can add a descriptive title, but if an issue is referenced, you don't have to.

## Checklist for submitter

- [ ] Check for conflict with integration test
- [ ] Unit tests pass

---

## Nominate for Group Code Review (Anyone can nominate it)
Indicate if you think the code should be reviewed in a Thursday code review session.

- [ ] Recommend for group code review

Also, nominate it on the code_review Slack channel.

### CODE REVIEW

File: src/adc_readout/AdcParse.cpp [hunk 1]
```diff
@@ -95,24 +95,50 @@ HeaderInfo parseHeader(const InData &Packet) {
...
+  if (Header.Type != PacketType::Idle and Header.Type != PacketType::Data) {
     throw ParserException(ParserException::Type::HEADER_TYPE);
   }
+  ReturnInfo.Type = Header.Type;
   ReturnInfo.DataStart = sizeof(PacketHeader);
   ReturnInfo.ReadoutCount = Header.ReadoutCount;
-  ReturnInfo.ReferenceTimestamp = Header.ReferenceTimeStamp;
+  ReturnInfo.ReferenceTimestamp = {Header.ReferenceTimeStamp,
+                                   TimeStamp::ClockMode(Header.ClockMode)};
+  if (Packet.Length != Header.ReadoutLength + 8u) {
```
Comment (mortenjc) [inline src/adc_readout/AdcParse.cpp]: + 8u? 

I assume this is some sort of padding, but it is probably better to make it explicit by using 
some well named variable.

File: src/adc_readout/AdcParse.h [hunk 2]
```diff
@@ -184,3 +199,14 @@ TrailerInfo parseTrailer(const InData &Packet, std::uint32_t StartByte);
...
+
+struct ConfigInfo {
+  enum class Version { VER_0, VER_1 } ProtocolVersion;
+  TimeStamp BaseTime;
+};
+
+/// \brief Extract information from the header of a packet for the purpose of
+/// setting up the parsing of subsequent packets. \param[in] Packet Raw data
+/// buffer. \return Packet information, i.e. protocol version and time stamp.
+/// \throw ParserException See exceptions type for possible parsing failures.
```
Comment (mortenjc) [inline src/adc_readout/AdcParse.h]: It is perhaps a little more readable if \param always appear at the beginning of the line?

File: src/adc_readout/AdcReadoutBase.cpp [hunk 3]
```diff
@@ -125,51 +126,59 @@ void AdcReadoutBase::packetFunction(InData const &Packet,
...
+    Receiver1->setPacketHandler(PacketHandler1);
+  };
+  auto GetPacketsForConfig = [&](auto Packet) {
+    try {
+      auto Config = parseHeaderForConfigInfo(Packet);
+      TimestampOffset = OffsetTime(ReadoutSettings.TimeOffsetSetting,
+                                   ReadoutSettings.ReferenceTime,
+                                   Config.BaseTime.getTimeStampNS());
+      LOG(INIT, Sev::Info, "Config packet received, starting data processing.");
+      SetUpInputForProcessing();
```
Comment (mortenjc) [inline src/adc_readout/AdcReadoutBase.cpp]: Couldyou not just do 
Receiver1->setPacketHandler(PacketHandler1);
here?

File: src/adc_readout/AdcReadoutConstants.h [hunk 4]
```diff
@@ -12,8 +12,6 @@
 #include <cstdint>
 
 /// \brief Clock frequency of the MRF timing hardware clock.
-static const std::int32_t TimerClockFrequency = 88052500;
-static const std::int32_t AdcSamplingRate = TimerClockFrequency / 2;
-static const std::int32_t AdcTimerCounterMax = AdcSamplingRate;
-static const double CycleLengthNS = 1e9 / TimerClockFrequency;
-static const double SampleLengthNS = 2e9 / TimerClockFrequency;
+static const std::int32_t TimerClockFrequencyExternal = 88052500;
```
Comment (mortenjc) [inline src/adc_readout/AdcReadoutConstants.h]: Maybe add // Hz just to be friendly

File: src/adc_readout/AdcSettings.cpp [hunk 5]
```diff
@@ -31,6 +60,41 @@ void setCLIArguments(CLI::App &Parser, AdcSettings &ReadoutSettings) {
                   "Kafka broker.")
       ->group("ADC Readout Options")
       ->default_str("AdcDemonstrator");
+
+  auto ParseOffsetTimestamp =
+      [&ReadoutSettings](std::vector<std::string> const &Input) -> bool {
+    auto TestString = Input.at(0);
+    std::transform(TestString.begin(), TestString.end(), TestString.begin(),
```
Comment (mortenjc) [inline src/adc_readout/AdcSettings.cpp]: Is there not a library function that can do this. Does boost have to_lower, should we put something like 
this helper into src/common ?

File: src/adc_readout/AdcTimeStamp.cpp [hunk 6]
```diff
@@ -11,50 +11,79 @@
...
 }
 
 // Note: This function might be significantly slower than CalcTimeStamp() for
 // some cases.
-std::uint64_t RawTimeStamp::getTimeStampNSFast() const {
+std::uint64_t TimeStamp::getTimeStampNSFast() const {
 
-  std::uint64_t NanoSec =
-      (((SecondsFrac * 100000000000) / (AdcTimerCounterMax)) + 50) / 100;
+  std::uint64_t NanoSec = (((CTime.SecondsFrac * 100000000000) /
```
Comment (mortenjc) [inline src/adc_readout/AdcTimeStamp.cpp]: A formula like this should maybe  have some explanatory text. 
Also what does the hardcoded 50 and 100 signify?

File: src/adc_readout/AdcParse.cpp (hunk 1 above)
Comment (SkyToGround) [reply src/adc_readout/AdcParse.cpp]: True, done.

File: src/adc_readout/AdcParse.cpp [hunk 7]
```diff
@@ -95,24 +95,50 @@ HeaderInfo parseHeader(const InData &Packet) {
...
+  if (Packet.Length < sizeof(PacketHeader)) {
+    throw ParserException(ParserException::Type::HEADER_LENGTH);
+  }
+  auto HeaderRaw = reinterpret_cast<const PacketHeader *>(Packet.Data);
+  PacketHeader Header(*HeaderRaw);
+  Header.fixEndian();
+  if (Header.Type != PacketType::Idle and Header.Type != PacketType::Data) {
+    throw ParserException(ParserException::Type::HEADER_TYPE);
+  }
+  std::map<Protocol, ConfigInfo::Version> VersionMap{
```
[the remark this answers is not in the document]
Comment (SkyToGround) [reply src/adc_readout/AdcParse.cpp]: This function is called once when processing the first packet.

File: src/adc_readout/AdcParse.h (hunk 2 above)
Comment (SkyToGround) [reply src/adc_readout/AdcParse.h]: I am pretty sure that this is caused by and old version of clang-format as I would never write this myself. I will fix it.

File: src/adc_readout/AdcReadoutBase.cpp (hunk 3 above)
Comment (SkyToGround) [reply src/adc_readout/AdcReadoutBase.cpp]: True, done.

File: src/adc_readout/AdcReadoutConstants.h (hunk 4 above)
Comment (SkyToGround) [reply src/adc_readout/AdcReadoutConstants.h]: Done.

File: src/adc_readout/AdcSettings.cpp (hunk 5 above)
Comment (SkyToGround) [reply src/adc_readout/AdcSettings.cpp]: 1. Weirdly, there is no function for doing this on a std::string in the standard library.
2. Yes, boost has a function that does this but I do not think it is worth it to add that dependency for this (relatively minor) functionality.
3. Yes, it *might* be worth it to add a function that does this into *common* but of course then we have the problems of: does the developers (i.e. us) know that it is there and how often do we use it assuming that we do remember.

File: src/adc_readout/AdcTimeStamp.cpp (hunk 6 above)
Comment (SkyToGround) [reply src/adc_readout/AdcTimeStamp.cpp]: This is a fixed point calculation of a decimal value, hence 50 corresponds to 0.5 and 100 makes it such that this is the case. Doing fixed point math can be significantly faster than floating point math but based on my benchmarks this is not necessarily the case on modern Intel CPUs that have powerful floating point units. I will hence remove the code.

**Pull request merged**